### PR TITLE
Login: Refactor show and hide

### PR DESCRIPTION
**Pull request by garrett**
Partially addresses #14828.

### CODE REVIEW

**Comment by garrett** [comment]
TODO:
- [ ] Convert all the showing and hiding to show() and hide()

**Comment by garrett** [comment]
I think I successfully changed everything over to hidden attribute showing and hiding with `show()` and `hide()` (and `hideToggle()`, which shows or hides based on a boolean) in JavaScript and `hidden` in the HTML in commit f8b0325.

Please review and test!

**Comment by garrett** [comment]
This refactoring was needed for #14688, to rip out arbitrary CSS in JavaScript, so I could properly implement the UI from the designs.

**Comment by mvollmer** [comment]
The "unsupported browser" state looks like this:

image

Shouldn't there be a message saying "Your browser is unsupported" or similar?  I check with master.

**Comment by mvollmer** [comment]
"OAuth logout" looks like this:

image

I check master as well. In any case, this doesn't look "designed". It has probably regressed over time.

**Comment by mvollmer** [comment]
Same for "Oauth failure":

image

**Comment by mvollmer** [comment]
> Shouldn't there be a message saying "Your browser is unsupported" or similar? I check with master.

Same on master -> #14884

**Comment by mvollmer** [comment]
> "OAuth logout" looks like this:

On master:

image

I think, this happens because `#login-again` gets `display: block` in the old code when it is shown.

File: src/ws/login.js [hunk 1]
```diff
@@ -74,21 +74,56 @@
...
             console.warn("fatal:", msg);
 
-        id("login-again").style.display = "none";
-        id("login-wait-validating").style.display = "none";
+        hide("#login-again", "#login-wait-validating");
 
         if (oauth_redirect_to) {
             id("login-again").href = oauth_redirect_to;
-            id("login-again").style.display = "block";
+            show("#login-again");
```
Comment (mvollmer) [inline src/ws/login.js:122]: `#login-again` is a `a` element, and giving it `display: block` had an effect. I think we should keep it `display: block` somehow, like with a explicit rule in login.css.

**Review by mvollmer — CHANGES_REQUESTED** [review]
Very nice!

File: src/ws/login.js [hunk 2]
```diff
@@ -506,26 +538,26 @@
...
-        id("login").style.visibility = 'visible';
-        id("login").style.display = "block";
-        id("user-group").style.display = in_conversation ? "none" : "block";
-        id("password-group").style.display = in_conversation ? "none" : "block";
-        id("conversation-group").style.display = in_conversation ? "block" : "none";
+        hide("#login-wait-validating");
+        show("#login");
+
+        hideToggle(["#user-group", "#password-group"], in_conversation);
+        hideToggle("#conservation-group", !in_conversation);
```
[the remark this answers is not in the document]
Comment (mvollmer) [reply src/ws/login.js:545]: This should explain most of the test failures.

File: src/ws/login.js (hunk 1 above)
Comment (mvollmer) [reply src/ws/login.js:122]: See https://github.com/cockpit-project/cockpit/pull/14870#issuecomment-723864991 and https://github.com/cockpit-project/cockpit/pull/14870#issuecomment-723878719

**Comment by garrett** [comment]
Rebased on current master; I'll look into and solve what I can now. I'll need some help with non-password logins. (Even though I guess they're currently broken in the same way on master too?)

File: src/ws/login.js (hunk 1 above)
Comment (garrett) [reply src/ws/login.js:122]: Yep. The display types should be in CSS, not JavaScript. I'll add it to the CSS.

**Comment by garrett** [comment]
> Shouldn't there be a message saying "Your browser is unsupported" or similar? I check with master.

Yes. It's:

```
          <h2 class="unsupported-browser-heading" translate>A modern browser is required for security, reliability, and performance.</h3>
```
(Oh, opens with h2, closed with h3?! I'll fix that too.)

Is this rewritten in the JavaScript perhaps?

**Comment by garrett** [comment]
Rebased, addressed concerns.

Specifically did not fix the browser message, as that's apparently broken on master as well. We should fix it, but probably in a follow-up, unless it's related to show/hide. (Which I don't think it is?)

**Comment by mvollmer** [comment]
> > Shouldn't there be a message saying "Your browser is unsupported" or similar? I check with master.
> 
> Yes. It's:
> 
> https://github.com/cockpit-project/cockpit/blob/4113c2e89f102cfc5f7291edd8676a675cf3adc4/src/ws/login.html#L109
> 
> Is this rewritten in the JavaScript perhaps?

No, it is moved off-screen by the CSS:

```
  .unsupported-browser-heading {
    font-size: 3rem;
    left: 50%;
    /* Pull the H2 half the width to the left to center */
    margin: 0 0 0 -32rem;
    position: absolute;
    text-align: center;
    top: -16rem;
    width: 64rem;
  }
```

Let's continue in  #14884.

**Review by mvollmer — APPROVED** [review]

**Comment by garrett** [comment]
> No, it is moved off-screen by the CSS

Oof. I wonder how that happened. That's definitely a hack. I guess that's from before I added grid.

**Comment by garrett** [comment]
I've made a draft @ #14891 to fix no JavaScript and the browser support messaging. (The support messaging fixups aren't there yet.)

But the important thing is that we're considering this PR done, as those aren't regressions compared to what's currently in our main branch — and these issues (no JS, unsupported browser) will be addressed in the follow-up.

**Pull request merged**